### ISSUE TITLE
New cuts cannot be used if they don't have a name

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#23]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/23) by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

```
#!python

ds = data.channel[11]
ds.cuts.cut(28, ds.p_pulse_rms[:]>3500)  # This does nothing
ds.cuts.cut(14, ds.p_pulse_rms[:]>3500)  # This applies the cut 

```

The latter works because cuts 0-14 have names. This leaves us no way to use cuts 15-31.

```
#!python


In [122]: print data.boolean_cut_desc
Out[122]: 
array([('pretrigger_rms', 1L), ('pretrigger_mean', 2L),
       ('pretrigger_mean_departure_from_median', 4L), ('peak_time_ms', 8L),
       ('rise_time_ms', 16L), ('postpeak_deriv', 32L),
       ('pulse_average', 64L), ('min_value', 128L),
       ('timestamp_sec', 256L), ('timestamp_diff_sec', 512L),
       ('peak_value', 1024L), ('energy', 2048L), ('timing', 4096L),
       ('p_filt_phase', 8192L), ('smart_cuts', 16384L), ('', 0L), ('', 0L),
       ('', 0L), ('', 0L), ('', 0L), ('', 0L), ('', 0L), ('', 0L),
       ('', 0L), ('', 0L), ('', 0L), ('', 0L), ('', 0L), ('', 0L),
       ('', 0L), ('', 0L), ('', 0L)], 
      dtype=[('name', 'S64'), ('mask', '<u4')])


```

The problem appears to be all those 0L values as the 2nd member of the tuples. @oneilg can watch this.

**Comment by joefowler** [comment]
**Original comment by Young Il Joe (Bitbucket: [jojo978](https://bitbucket.org/jojo978), GitHub: [jojo978](https://github.com/jojo978)).**

----------------------------------------

In order to use non-builtin cut, you need to register first.
It's reasonable to call that boolean cut field "pulse_rms", so you can register like this. Then you can use the name of this cut field.

```
#!python

data.register_boolean_cut_fields("pulse_rms")
ds.cuts.cut("pulse_rms", ds.p_pulse_rms[:] > 3500)
ds.cuts.cut(16, ds.p_pulse_rms[:] > 3500)  # this might work too.
```

cut_num of "pulse_rms" field would not be 15 because "calibration" categorical field takes bit #15. If "pulse_rms" is the first non-buildin cut field, its cut_num will be 16.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

As Young said, it is intended that you never use a cut number directly, instead always naming your cuts. The number syntax exists for backwards compatibility. It should not fail silently however.

This is mainly a documentation (mass documentation?? it could happen) issue.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Well, this is a reasonable design choice, but it breaks every existing script that *does* use a cut number (i.e., every special-purpose cut I have ever made).

So...apart from any so-called *documentation*, we need the use of a cut-by-number to fail with an explicitly raised error instead of silently just not cutting anything.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Agreed. I guess it could work instead of throwing an error. I think we will be better off moving people to use the API for named cuts by fast by having it error out. We all agree failing silently is not acceptable. 

Check out my first attempt at documentation 

https://bitbucket.org/joe_fowler/mass/src/3c2f19e866da7e641d7389b21f4bdff63d3e7df5/doc/multibitcutsdoc.md?at=master&fileviewer=file-view-default

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Okay. We can use cuts-by-name only, but don't close this issue until an explicit error is raised when someone tries to cut-by-number.

**Comment by joefowler** [comment]
**Original comment by Young Il Joe (Bitbucket: [jojo978](https://bitbucket.org/jojo978), GitHub: [jojo978](https://github.com/jojo978)).**

----------------------------------------

You can use cut_num for ds.cuts.cut. And In the setup_cleanup branch, it will raise an exception when you use an unregistered cut_num (not operating with a 0x0000 mask (because it's is the bit mask of unregister boolean fields.), which means it doesn't anything.).

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Young fixed this in ce800784
